src/query/vacuum.c

```diff
@@ -5629,34 +5629,18 @@ vacuum_update_keep_from_log_pageid (THREAD_ENTRY * thread_p)
    * does not remove log required for vacuum.
    * If vacuum data is empty, then all blocks until (and including) vacuum_Data.last_blockid have been
    * vacuumed, and first page belonging to next block must be preserved (this is most likely in the active area of the
-   * log, for now). However, it might happen that the page referred might belong in a log archive that might have
-   * been removed due to a previous action. So to be sure, we set the pageid, from which the vacuum must keep
-   * the remaining pages, to NULL_PAGEID.
+   * log, but not always).
    * If vacuum data is not empty, then we need to preserve the log starting with the first page of first unvacuumed
    * block.
    */
-  VACUUM_LOG_BLOCKID keep_from_blockid;
-
   if (vacuum_is_empty ())
     {
-      LOG_LSA last_mvcc_lsa = log_Gl.hdr.mvcc_op_log_lsa;
-      if (last_mvcc_lsa.is_null () || vacuum_get_log_blockid (last_mvcc_lsa.pageid) <= vacuum_Data.get_last_blockid ())
-	{
-	  /* safe to remove all archives */
-	  keep_from_blockid = VACUUM_NULL_LOG_BLOCKID;
-	  vacuum_Data.keep_from_log_pageid = NULL_PAGEID;
-	}
-      else
-	{
-	  /* keep block of log_Gl.hdr.mvcc_op_log_lsa */
-	  keep_from_blockid = vacuum_get_log_blockid (log_Gl.hdr.mvcc_op_log_lsa.pageid);
-	  vacuum_Data.keep_from_log_pageid = VACUUM_FIRST_LOG_PAGEID_IN_BLOCK (keep_from_blockid);
-	}
+      // keep starting with next after last_blockid ()
```

Comment:
What if get_last_block_id modifies between vacuum_is_empty and vacuum_Data.keep_from_log_pageid =... I think in this case we are in trouble.

Reply:
only vacuum master manipulates vacuum data fields. last blockid / empty data status cannot change.